### PR TITLE
[Merged by Bors] - feat(KrullDimension): concrete calculations

### PR DESCRIPTION
**Pull request by nomeata**
Part of https://github.com/leanprover-community/mathlib4/pull/15524. This adds concrete calculations of the height and Krull Dimension for Nat, Int, WithTop, WithBot and ENat.

From the Carleson project.

---

* [x] depends on: #19251

### CODE REVIEW

File: Mathlib/Order/KrullDimension.lean [hunk 1]
```diff
@@ -616,4 +617,138 @@ lemma coheight_bot_eq_krullDim [OrderBot α] : coheight (⊥ : α) = krullDim α
...
+  apply le_antisymm
+  · apply height_le_coe_iff.mpr
+    simp +contextual only [ih, Nat.cast_lt, implies_true]
+  · exact length_le_height_last (p := LTSeries.range n)
+
+@[simp] lemma coheight_nat (n : ℕ) : coheight n = ⊤ := by
+  apply coheight_eq_top_iff.mpr
+  intro m
+  use (LTSeries.range m).map (· + n) (StrictMono.add_const (fun _ _ x ↦ x) n)
+  simp
```
[the remark this answers is not in the document]
Comment (nomeata) [reply Mathlib/Order/KrullDimension.lean:639]: I'll give it a try. Should it then still be a simp lemma, or is that not good given the then rather unspecific lhs?

[the remark this answers is not in the document]
Comment (nomeata) [reply Mathlib/Order/KrullDimension.lean]: I gave it a try, it works. It seems I have to add
```
instance Int.instNoMinOrder : NoMinOrder ℤ where
  exists_lt n := ⟨n - 1, n.pred_self_lt⟩
```
though – where should I put them?

[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Order/KrullDimension.lean]: Are you sure this is missing? What does
```
import Mathlib

#synth NoMinOrder ℤ 
```
say?

If it really is missing, it should go in `Order.Int`

[the remark this answers is not in the document]
Comment (nomeata) [reply Mathlib/Order/KrullDimension.lean]: Oh, my bad, I looked through the instances at https://loogle.lean-lang.org/?q=%7C-+NoMinOrder+_ but didn’t expect a level of indirection. But that's how mathlib works :-)

File: Mathlib/Data/Int/Order/Basic.lean [hunk 2]
```diff
@@ -66,4 +66,5 @@ theorem mul_nonneg_of_nonneg_or_nonpos {a b : ℤ} : 0 ≤ a ∧ 0 ≤ b ∨ a 
 
 protected theorem mul_nonneg_iff {a b : ℤ} : 0 ≤ a * b ↔ 0 ≤ a ∧ 0 ≤ b ∨ a ≤ 0 ∧ b ≤ 0 :=
   ⟨nonneg_or_nonpos_of_mul_nonneg, mul_nonneg_of_nonneg_or_nonpos⟩
+
```
Comment (nomeata) [inline Mathlib/Data/Int/Order/Basic.lean:69]:
```suggestion
```

File: Mathlib/Order/KrullDimension.lean [hunk 3]
```diff
@@ -616,4 +617,149 @@ lemma coheight_bot_eq_krullDim [OrderBot α] : coheight (⊥ : α) = krullDim α
...
+  simp [krullDim_eq_iSup_coheight, coheight_of_noMaxOrder]
+
+@[simp] lemma krullDim_of_noMinOrder [Nonempty α] [NoMinOrder α] : krullDim α = ⊤ := by
+  simp [krullDim_eq_iSup_height, height_of_noMinOrder]
+
+lemma coheight_nat (n : ℕ) : coheight n = ⊤ := coheight_of_noMaxOrder ..
+
+lemma krullDim_nat : krullDim ℕ = ⊤ := krullDim_of_noMaxOrder ..
+
+lemma height_int (n : ℤ) : height n = ⊤ := height_of_noMinOrder ..
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Order/KrullDimension.lean:664]: I think they are fine to have

File: Mathlib/Order/KrullDimension.lean [hunk 4]
```diff
@@ -7,6 +7,7 @@ Authors: Jujian Zhang, Fangming Li, Joachim Breitner
 import Mathlib.Order.RelSeries
 import Mathlib.Order.Minimal
 import Mathlib.Data.ENat.Lattice
+import Mathlib.Algebra.Order.Group.Int
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:10]: Can you keep this alphabetical?

File: Mathlib/Order/KrullDimension.lean [hunk 5]
```diff
@@ -616,4 +617,149 @@ lemma coheight_bot_eq_krullDim [OrderBot α] : coheight (⊥ : α) = krullDim α
...
+  apply Equiv.iSup_congr e
+  simp [e]
+
+@[simp] lemma krullDim_withBot [Nonempty α] : krullDim (WithBot α) = krullDim α + 1 := by
+  conv_lhs => rw [← krullDim_orderDual]
+  conv_rhs => rw [← krullDim_orderDual]
+  exact krullDim_WithTop (α := αᵒᵈ)
+
+@[simp]
+lemma krullDim_ENat : krullDim ℕ∞ = ⊤ := by
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:747]:
```suggestion
lemma krullDim_enat : krullDim ℕ∞ = ⊤ := by
```
same below

**Review by YaelDillies — COMMENTED** [review]
I think a bunch of these would be easier if you first proved that in a linear order one has `height a = (Set.Iio a).ncard` and `coheight a = (Set.Ioi a).ncard`

File: Mathlib/Order/KrullDimension.lean [hunk 6]
```diff
@@ -616,4 +617,138 @@ lemma coheight_bot_eq_krullDim [OrderBot α] : coheight (⊥ : α) = krullDim α
...
+@[simp] lemma krullDim_WithTop [Nonempty α] : krullDim (WithTop α) = krullDim α + 1 := by
+  rw [← height_top_eq_krullDim, krullDim_eq_iSup_height_of_nonempty, height_eq_iSup_lt_height]
+  norm_cast
+  rw [ENat.iSup_add, iSup_subtype']
+  symm
+  let e : α ≃ {y : WithTop α // y < ⊤} :=
+    { toFun := fun x => ⟨x, WithTop.coe_lt_top x⟩
+      invFun := fun ⟨x,h⟩ => WithTop.untop x (ne_top_of_lt h)
+      left_inv := fun _ => by simp,
+      right_inv := fun _ => by simp }
```
[the remark this answers is not in the document]
Comment (nomeata) [reply Mathlib/Order/KrullDimension.lean]: https://github.com/leanprover-community/mathlib4/pull/19251

File: Mathlib/Order/WithBot.lean [hunk 7]
```diff
@@ -158,6 +158,13 @@ theorem eq_unbot_iff {a : α} {b : WithBot α} (h : b ≠ ⊥) :
   · simpa using h rfl
   · simp
 
+/-- The equivalence between the non-bottom elements of `WithBot α` and `α`. -/
+def subtype_ne_bot_equiv : {y : WithBot α // y ≠ ⊥} ≃ α where
```
Comment (YaelDillies) [inline Mathlib/Order/WithBot.lean:162]: Can you camel case?

File: Mathlib/Order/WithBot.lean [hunk 8]
```diff
@@ -158,6 +158,13 @@ theorem eq_unbot_iff {a : α} {b : WithBot α} (h : b ≠ ⊥) :
...
+  toFun := fun ⟨x,h⟩ => WithBot.unbot x h
+  invFun := fun x => ⟨x, WithBot.coe_ne_bot⟩
+  left_inv := fun _ => by simp
+  right_inv := fun _ => by simp
```
[the remark this answers is not in the document]
Comment (nomeata) [reply Mathlib/Order/WithBot.lean:166]: (Applied in #19251.)

**Comment by nomeata** [comment]
Here is a sketch of the `LinearOrder` proof:
```
theorem height_of_linearOrder {α : Type*} [LinearOrder α] (a : α) :
    height a = (Set.Iio a).encard := by
  apply le_antisymm
  · apply height_le
    intro p hlast
    calc ↑p.length
      _ = (Set.Iio p.length).encard := by
        rw [← Set.Finite.cast_ncard_eq (by toFinite_tac)]
        rw [Set.ncard_eq_toFinset_card _]
        simp
      _ ≤ (Set.Iio a).encard := by
        apply Set.encard_le_encard_of_injOn (f := fun (i : Nat) => p i)
        case hf =>
          intro i hi
          simp at hi
          simp only [Set.mem_Iio, ←hlast]
          apply p.rel_of_lt
          simp [Fin.last]
          apply Fin.lt_def.mpr
          simp
          rw [Nat.mod_eq_of_lt (by omega)]
          exact hi
        case f_inj =>
          apply StrictMonoOn.injOn
          intro i hi j hj hij
          simp at hi hj
          apply p.rel_of_lt
          apply Fin.lt_def.mpr
          simp only [Fin.val_natCast]
          rw [Nat.mod_eq_of_lt (by omega)]
          rw [Nat.mod_eq_of_lt (by omega)]
          exact hij
  · cases hc : (Set.Iio a).encard with
    | top =>
      simp only [top_le_iff]
      apply height_eq_top_iff.mpr
      intro n
      have p : LTSeries α := sorry
      have hlen : p.length = n := sorry
      have hlast : p.last = a := sorry
      exact ⟨p, hlast, hlen⟩
    | coe n =>
      replace hc := congrArg ENat.toNat hc
      simp [← Set.ncard_def] at hc
      have p : LTSeries α := sorry
      have hlen : p.length = n := sorry
      have hlast : p.last = a := sorry
      rw [← hlen, ← hlast]
      apply length_le_height_last

**Review by YaelDillies — APPROVED** [review]
Okay, looks complicated enough to be postponed.

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**